src/Compilers/Core/Portable/SourceGeneration/GeneratorDriver.cs

```diff
@@ -46,80 +46,19 @@ internal GeneratorDriver(ParseOptions parseOptions, ImmutableArray<ISourceGenera
-
-                // initialize the generator if needed
-                if (!generatorState.Info.Initialized)
-                {
-                    generatorState = InitializeGenerator(generator, diagnosticsBag, cancellationToken);
-                }
-                stateBuilder.Add(generatorState);
-
-                // create the syntax receiver if requested
-                if (generatorState.Info.SyntaxReceiverCreator is object)
-                {
-                    var rx = generatorState.Info.SyntaxReceiverCreator();
-                    receivers.Add(generator, rx);
-                }
-            }
-
-            // Run a syntax walk if any of the generators requested it
-            if (receivers.Count > 0)
-            {
-                GeneratorSyntaxWalker walker = new GeneratorSyntaxWalker(receivers.Values.ToImmutableArray());
-                foreach (var syntaxTree in compilation.SyntaxTrees)
-                {
-                    walker.Visit(syntaxTree.GetRoot(cancellationToken));
-                }
-            }
-
-            // https://github.com/dotnet/roslyn/issues/42629: should be possible to parallelize this
-            for (int i = 0; i < state.Generators.Length; i++)
-            {
-                var generator = state.Generators[i];
-                var generatorState = stateBuilder[i];
-                try
-                {
-                    // don't try and generate if initialization failed
-                    if (!generatorState.Info.Initialized)
-                    {
-                        continue;
-                    }
+            var state = RunGeneratorsCore(compilation, new DiagnosticBag(), cancellationToken);
+            return FromState(state);
+        }
 
-                    // we create a new context for each run of the generator. We'll never re-use existing state, only replace anything we have
-                    _ = receivers.TryGetValue(generator, out var syntaxReceiverOpt);
-                    var context = new SourceGeneratorContext(compilation, state.AdditionalTexts.NullToEmpty(), state.OptionsProvider, syntaxReceiverOpt, diagnosticsBag);
-                    generator.Execute(context);
-                    stateBuilder[i] = generatorState.WithSources(ParseAdditionalSources(generator, context.AdditionalSources.ToImmutableAndFree(), cancellationToken));
-                }
-                catch
-                {
-                    diagnosticsBag.Add(Diagnostic.Create(MessageProvider, MessageProvider.WRN_GeneratorFailedDuringGeneration, generator.GetType().Name));
-                }
-            }
-            state = state.With(generatorStates: stateBuilder.ToImmutableAndFree());
-            diagnostics = diagnosticsBag.ToReadOnlyAndFree();
+        public GeneratorDriver RunFullGeneration(Compilation compilation, out Compilation outputCompilation, out ImmutableArray<Diagnostic> diagnostics, CancellationToken cancellationToken = default)
```

Comment:
Do you imagine this API going away, or you still want to keep it around for the "easy" case?

Reply:
Hmm, I think it's super useful for unit testing, self host scenarios etc. I'm conflicted though because of the whole 'never remove an API' thing.

---
In reply to: [465386510](https://github.com/dotnet/roslyn/pull/46296#discussion_r465386510) [](ancestors = 465386510)